CRM/Member/BAO/Membership.php

```diff
@@ -1603,17 +1580,33 @@ public static function buildMembershipTypeValues(&$form, $membershipTypeID = arr
       'max_related',
       'duration_unit',
       'duration_interval',
-    );
+    ];
 
-    while ($dao->fetch()) {
-      $membershipTypeValues[$dao->id] = array();
-      foreach ($membershipTypeFields as $mtField) {
-        $membershipTypeValues[$dao->id][$mtField] = $dao->$mtField;
-      }
+    $membershipTypeParams = [
+      'return' => $membershipTypeFields,
+      'options' => ['limit' => 0],
+    ];
+    if ($activeOnly) {
+      $membershipTypeParams['is_active'] = 1;
+    }
+    if ($financialTypes) {
+      $membershipTypeParams['financial_type_id'] = ['IN' => array_keys($financialTypes)];
+    }
+    if (!empty($membershipTypeIDS)) {
+      $membershipTypeParams['id'] = ['IN' => $membershipTypeIDS];
     }
 
+    $membershipTypesApi = civicrm_api3('MembershipType', 'get', $membershipTypeParams);
+    $membershipTypeValues = CRM_Utils_Array::value('values', $membershipTypesApi);
+    // @fixme The foreach to modify the API values is just to make sure we return exactly the same as this function did before switching to the API
+    foreach ($membershipTypeValues as $fmembershipTypeID => $membershipTypeDetail) {
+      $membershipTypeValues[$fmembershipTypeID]['relationship_type_id'] = isset($membershipTypeDetail['relationship_type_id']) ? CRM_Utils_Array::first($membershipTypeDetail['relationship_type_id']) : NULL;
```

Comment:
I think your goal here is no- change - ie. don't go from a string to an array - but it is a DIFFERENT string - ie if there are multiple then before it would be a separated list whereas after it would just be one.